### ISSUE TITLE
`memeq20` seems to actually comapre 24 bytes, while `mass_comparison` only 12

### ISSUE THREAD

**Issue by travisdowns** [lemire/Code-used-on-Daniel-Lemire-s-blog#21]
Unless I'm misunderstanding, the `memeq20` function seems to compare 8 * 3 == 24 bytes:

```
bool memeq20(const char * s1, const char * s2) {
    uint64_t w1, w2;
    memcpy(&w1, s1, sizeof(w1));
    memcpy(&w2, s2, sizeof(w2));
    bool answer1 = (w1 == w2);
    memcpy(&w1, s1 + sizeof(w1), sizeof(w1));
    memcpy(&w2, s2 + sizeof(w2), sizeof(w2));
    bool answer2 = (w1 == w2);
    uint64_t ww1, ww2;
    memcpy(&ww1, s1 + 2 * sizeof(w1), sizeof(ww1));
    memcpy(&ww2, s2 + 2 * sizeof(w1), sizeof(ww2));
    bool answer3 = (ww1 == ww2);
    return answer1 && answer2 && answer3;
}
```

Each `==` is between 64-bit values and they are all apparently of distinct bytes.

Conversely, we have `#define STRINGLENGTH 12` so I think `mass_comparison` is only comparing 12 bytes.

**Comment by lemire** [comment]
Thank you. I will push a fix.

**Comment by lemire** [comment]
Note: it does not really impact the blog post.

**Comment by lemire** [comment]
>In practice it depends heavily on the compiler and other factors.

I agree with that, see this paragraph in my blog post...

`I am using GCC 5.5. Your results will vary quite a bit depending on the compiler. Also, there might be branching involved, so the results will depend on the statistics of your data.`


My code is not smart... and I would not be surprised if a compiler could generate better code with memcmp than my equality-checking value... but it is not the case in general.

**Comment by travisdowns** [comment]
> but it is not the case in general.

What do you mean by that?

That it is not the case on the compiler(s) you tried that they generate better code?

That if everything was done ideally lexo compare is fundamentally more ineffiecient, so even if they generate better code than your equals code, it is only because the equals code is not particularly efficient (actually I think it is)?

Something else?

**Comment by travisdowns** [comment]
> but it is not the case in general

Right, but by this you mean that the lexicographic compare can _at best_ tie the pure equality approach, which is very different than saying the lexicographic approach is slower (due to the tie case).

The million dollar question of course is whether it ties or is very close in practice. Actually for some implementations of equality checking I think it is very close or tied and for long comparisons it is very close. There are some short match cases where equality is a fair amount better - but the test isn't showing this because of the impact of library call vs inlined non-library call. It is measuring something completely different.

Big-endian machines have an inherent advantage here because the bytes are already in the right order for the "subtract" trick when you find the mismatched word. On x86 you need a bswap, which slows things down a bit.
